README
Net-CIDR-Lookup

This module provides a pure-Perl implementation of a tree of netblocks. It
installs like most other modules (see generic instructions below) and depends
only on Test::Class for the unit tests (which you can skip if you like).


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Net::CIDR::Lookup

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=CIDR-Lookup

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/CIDR-Lookup

    CPAN Ratings
        http://cpanratings.perl.org/d/CIDR-Lookup

    Search CPAN
        http://search.cpan.org/dist/CIDR-Lookup/


COPYRIGHT AND LICENCE

Copyright (C) 2008, 2012 Matthias Bethke <matthias@towiski.de>

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for Net::CIDR::Lookup

0.3     Thu Oct 30 13:12:08 CET 2008
        First version released to CPAN



t/lib/Net/CIDR/Lookup/Test.pm
package Net::CIDR::Lookup::Test;

use strict;
use warnings;

use base 'Test::Class';
use Test::More;
use Net::CIDR::Lookup;

#-------------------------------------------------------------------------------

sub check_methods : Test(startup => 8) {
    my $t = Net::CIDR::Lookup->new;
    can_ok($t,'add');
    can_ok($t,'add_num');
    can_ok($t,'add_range');
    can_ok($t,'lookup');
    can_ok($t,'lookup_num');
    can_ok($t,'clear');
    can_ok($t,'dump');
    can_ok($t,'walk');
}

sub before : Test(setup) {
    my $self = shift;
    $self->{tree} = Net::CIDR::Lookup->new;
}

#-------------------------------------------------------------------------------

sub add : Tests(5) {
    my $self = shift;
    my $t = $self->{tree};
    is($t->add('192.168.0.129/25', 42), 1, 'add() succeeded');
    is($t->add('1.2.0.0/15', 23), 1, 'add() succeeded');
    is($t->lookup('192.168.0.161'), 42, 'Block 192.168.0.129/25 lookup OK');
    is($t->lookup('1.3.123.234'), 23, 'Block 1.2.0.0/15 lookup OK');
    is($t->lookup('2.3.4.5'), undef, 'No result outside blocks');
}

sub add_range : Tests(6) {
    my $self = shift;
    my $t = $self->{tree};
    is($t->add_range('192.168.0.130-192.170.0.1', 42), 1, 'add_range() succeeded');
    is($t->add_range('1.3.123.234 - 1.3.123.240', 23), 1, 'add_range() succeeded');
    is($t->lookup('192.169.0.22'), 42, 'Range 192.168.0.130 - 192.170.0.1');
    is($t->lookup('1.3.123.235'),  23, 'Range 1.3.123.234 - 1.3.123.240');
    is($t->lookup('2.3.4.5'), undef, 'No result outside blocks');
    my $h = $t->dump;
    is(scalar keys %$h, 19, 'Range expansion: number of keys');
}

sub collision : Tests(1) {
    my $self = shift;
    my $t = $self->{tree};
    $t->add('192.168.0.129/25', 42);
    isnt($t->add('192.168.0.160/31', 23), 1, 'Collision: add() failed as expected');
    # TODO errstr?
}

sub benign_collision : Test(1){
    my $self = shift;
    my $t = $self->{tree};
    $t->add('192.168.0.129/25', 42);
    is($t->add('192.168.0.160/31', 42), 1, 'Benign collision: add() succeeded');
}

sub merger : Tests(3) {
    my $self = shift;
    my $t = $self->{tree};
    $t->add('192.168.0.130/25', 42);
    is($t->add('192.168.0.0/25', 42),   1, 'Merger: add() succeeded');
    my $h = $t->dump;
    is(scalar keys %$h, 1, 'Merged block: number of keys');
    my ($k,$v) = each %$h;
    is($k, '192.168.0.0/24', 'Merged block: correct merged net block');
}

sub recursive_merger : Tests(4) {
    my $self = shift;
    my $t = $self->{tree};
    $t->add('0.1.1.0/24', 42);
    is($t->add('0.1.0.128/25', 42), 1, 'Recursive merger: add(0.0.0.128/25) succeeded');
    is($t->add('0.1.0.0/25', 42),   1, 'Recursive merger: add(0.0.0.0/25) succeeded');
    my $h = $t->dump;
    is(scalar keys %$h, 1, 'Recursively merged block: number of keys');
    my ($k,$v) = each %$h;
    is($k, '0.1.0.0/23', 'Recursively merged block: correct merged net block');
}

sub nonmerger : Tests(2) {
    my $self = shift;
    my $t = $self->{tree};
    $t->add('192.168.0.130/25', 42);
    is($t->add('192.168.0.0/25', 23),   1, 'Non-merger: add() succeeded');
    my $h = $t->dump;
    is(scalar keys %$h, 2, 'Unmerged adjacent blocks: correct number of keys');
}

sub equalrange : Tests(3) {
    my $self = shift;
    my $t = $self->{tree};
    $t->add('192.168.0.130/25', 1);
    is($t->add('192.168.0.130/25', 1), 1, 'Equalrange: add() succeeded');
    my $h = $t->dump;
    is(0+keys %$h, 1, 'Got single block from two equal inserts');
    is($h->{'192.168.0.128/25'}, 1, 'Got correct block');
}

sub subrange1 : Tests(3) {
    my $self = shift;
    my $t = $self->{tree};
    $t->add('192.168.0.1/24', 1);
    is($t->add('192.168.0.1/25', 1), 1, 'Immediate subrange: add() succeeded');
    my $h = $t->dump;
    is(0+keys %$h, 1, 'Immediate subrange: resulted in single block');
    is($h->{'192.168.0.0/24'}, 1, 'Immediate subrange: got correct block');
}

sub subrange2 : Tests(3) {
    my $self = shift;
    my $t = $self->{tree};
    $t->add('192.168.0.1/24', 1);
    is($t->add('192.168.0.1/28', 1), 1, 'Small subrange: add() succeeded');
    my $h = $t->dump;
    is(0+keys %$h, 1, 'Small subrange: resulted in single block');
    is($h->{'192.168.0.0/24'}, 1, 'Small subrange: got correct block');
}

sub superrange1 : Tests(3) {
    my $self = shift;
    my $t = $self->{tree};
    $t->add('192.168.0.128/25', 1);
    is($t->add('192.168.0.0/24', 1), 1,   'Immediate superrange: add() succeeded');
    my $h = $t->dump;
    is(0+keys %$h, 1, 'Immediate superrange: resulted in single block');
    is($h->{'192.168.0.0/24'}, 1, 'Immediate superrange: got correct block');
}

sub superrange2 : Tests(3) {
    my $self = shift;
    my $t = $self->{tree};
    $t->add('192.168.160.128/25', 1);
    is($t->add('192.168.160.0/20', 1), 1, 'Big superrange: add() succeeded');
    my $h = $t->dump;
    is(0+keys %$h, 1, 'Big superrange: resulted in single block');
    is($h->{'192.168.160.0/20'}, 1, 'Big superrange: got correct block');
}

sub clear : Tests(1) {
    my $self = shift;
    my $t = $self->{tree};
    $t->add('192.168.0.129/25', 42);
    $t->clear;
    is(scalar keys %{$t->dump}, 0, 'Reinitialized tree');
}

1;



t/lib/Net/CIDR/Lookup/IPv6/Test.pm
package Net::CIDR::Lookup::IPv6::Test;

use strict;
use warnings;

use base 'Test::Class';
use Test::More;
use Test::Exception;
use Net::CIDR::Lookup::IPv6;

#-------------------------------------------------------------------------------

sub check_methods : Test(startup => 8) {
    my $t = Net::CIDR::Lookup::IPv6->new;
    can_ok($t,'add');
    can_ok($t,'add_num');
    can_ok($t,'add_range');
    can_ok($t,'lookup');
    can_ok($t,'lookup_num');
    can_ok($t,'clear');
    can_ok($t,'dump');
    can_ok($t,'walk');
}

sub before : Test(setup) {
    my $self = shift;
    $self->{tree} = Net::CIDR::Lookup::IPv6->new;
}

sub _needs_ipv6 : Test(setup) {
    shift->SKIP_ALL('needs IPv6 support in Socket module')
    unless defined eval { Socket::AF_INET6() };
}
#-------------------------------------------------------------------------------

sub add : Tests(5) {
    my $self = shift;
    my $t = $self->{tree};
    lives_ok(sub { $t->add('2001:db8::/32', 42) }, 'add() succeeded');
    lives_ok(sub { $t->add('2002:db8::/31', 23) }, 'add() succeeded');
    is($t->lookup('2001:db8::1234'), 42, 'Block 2001:db8::/32 lookup OK');
    is($t->lookup('2002:db8:1::'), 23, 'Block 2002:db8::/31 lookup OK');
    is($t->lookup('::1'), undef, 'No result outside blocks');
}

sub add_range : Tests(4) {
    my $self = shift;
    my $t = $self->{tree};
    $t->add_range('2001:db8::-2003:db8::abc', 42);
    $t->add_range('1::1234 - 1::1:2345', 23);
    is($t->lookup('2002:cb8::abc'),  42, 'Range 2001:db8::--2002:db8::abc OK');
    is($t->lookup('1::ffff'), 23, 'Range 1::1234--1::1:2345 OK');
    is($t->lookup('f::'), undef, 'No result outside blocks');
    my $h = $t->dump;
    is(scalar keys %$h, 39, 'Range expansion: number of keys');
}

1;
__END__

sub collision : Tests(2) {
    my $self = shift;
    my $t = $self->{tree};
    lives_ok(sub { $t->add('192.168.0.129/25', 42) }, 'add() succeeded');
    isnt($t->add('192.168.0.160/31', 23), 1, 'add() failed as expected');
    # TODO errstr?
}

sub benign_collision : Test(3){
    my $self = shift;
    my $t = $self->{tree};
    lives_ok(sub { $t->add('192.168.0.129/25', 42) }, 'add() succeeded');
    lives_ok(sub { $t->add('192.168.0.160/31', 42) }, 'add() succeeded');
    is($Net::CIDR::Lookup::errstr, undef, 'Benign CIDR block collision');
}

sub merger : Tests(4) {
    my $self = shift;
    my $t = $self->{tree};
    lives_ok(sub { $t->add('192.168.0.130/25', 42) }, 'add() succeeded');
    is($t->add('192.168.0.0/25', 42),   1, 'add() succeeded');
    my $h = $t->dump;
    is(scalar keys %$h, 1, 'Merged block: number of keys');
    my ($k,$v) = each %$h;
    is($k, '192.168.0.0/24', 'Merged block: correct merged net block');
}

sub nonmerger : Tests(3) {
    my $self = shift;
    my $t = $self->{tree};
    lives_ok(sub { $t->add('192.168.0.130/25', 42) }, 'add() succeeded');
    is($t->add('192.168.0.0/25', 23),   1, 'add() succeeded');
    my $h = $t->dump;
    is(scalar keys %$h, 2, 'Unmerged adjacent blocks: correct number of keys');
}

sub equalrange : Tests(4) {
    my $self = shift;
    my $t = $self->{tree};
    lives_ok(sub { $t->add('192.168.0.130/25', 1) }, 'add() succeeded');
    lives_ok(sub { $t->add('192.168.0.130/25', 1) }, 'add() succeeded');
    my $h = $t->dump;
    is(0+keys %$h, 1, 'Got single block from two equal inserts');
    is($h->{'192.168.0.128/25'}, 1, 'Got correct block');
}

sub subrange1 : Tests(4) {
    my $self = shift;
    my $t = $self->{tree};
    lives_ok(sub { $t->add('192.168.0.1/24', 1) }, 'add() range succeeded');
    lives_ok(sub { $t->add('192.168.0.1/25', 1) }, 'add() subrange succeeded');
    my $h = $t->dump;
    is(0+keys %$h, 1, 'Got single block from range followed by one of its immediate subranges');
    is($h->{'192.168.0.0/24'}, 1, 'Got correct block');
}

sub subrange2 : Tests(4) {
    my $self = shift;
    my $t = $self->{tree};
    lives_ok(sub { $t->add('192.168.0.1/24', 1) }, 'add() range succeeded');
    lives_ok(sub { $t->add('192.168.0.1/28', 1) }, 'add() subrange succeeded');
    my $h = $t->dump;
    is(0+keys %$h, 1, 'Got single block from range followed by a small subrange');
    is($h->{'192.168.0.0/24'}, 1, 'Got correct block');
}

sub superrange1 : Tests(4) {
    my $self = shift;
    my $t = $self->{tree};
    lives_ok(sub { $t->add('192.168.0.128/25', 1) }, 'add() range succeeded');
    is($t->add('192.168.0.0/24', 1), 1,   'add() superrange succeeded');
    my $h = $t->dump;
    is(0+keys %$h, 1, 'Got single block from range followed by its superrange');
    is($h->{'192.168.0.0/24'}, 1, 'Got correct block');
}

sub superrange2 : Tests(4) {
    my $self = shift;
    my $t = $self->{tree};
    lives_ok(sub { $t->add('192.168.160.128/25', 1) }, 'add() range succeeded');
    lives_ok(sub { $t->add('192.168.160.0/20', 1) }, 'add() superrange succeeded');
    my $h = $t->dump;
    is(0+keys %$h, 1, 'Got single block from range followed by one of its higher superranges');
    is($h->{'192.168.160.0/20'}, 1, 'Got correct block');
}

sub clear : Tests(2) {
    my $self = shift;
    my $t = $self->{tree};
    lives_ok(sub { $t->add('192.168.0.129/25', 42) }, 'add() succeeded');
    $t->clear;
    is(scalar keys %{$t->dump}, 0, 'Reinitialized tree');
}

1;



lib/Net/CIDR/Lookup/Tie.pm
=head1 NAME

Net::CIDR::Lookup::Tie

=head1 DESCRIPTION

This is a L<Tie::Hash> interface to L<Net::CIDR::Lookup>, see there for
details.

The tied hash accepts net blocks as keys in the same syntax as
C<Net::CIDR::Lookup->add()> or C<add_range()> and stores arbitrary (with the
exception of C<undef>) scalar values under these. The same coalescing as in
C<Net::CIDR::Lookup> takes place, so if you add any number of different keys
you may end up with a hash containing I<less> keys if any mergers took place.

Of course you can retrieve the corresponding net block's value for any key that
is I<contained> within that block, so the number of possible lookup keys is
usually far greater than that of explicitly stored key/value pairs.

=head1 SYNOPSIS

  use Net::CIDR::Lookup::Tie;

  tie my %t, 'Net::CIDR::Lookup::Tie';
  $t{'192.168.42.0/24'}   = 1;    # Add first network
  $t{'192.168.43.0/24'}   = 1;    # Automatic coalescing to a /23
  $t{'192.168.41.0/24'}   = 2;    # Stays separate due to different value

  print $t{'192.168.42.100'};		 # prints "1"

  foreach(keys %h) { ... }			 # Do anything you'd do with a regular hash

=head1 HISTORY

=over 1

=item v0.3 First CPAN release

=item v0.41 Version bumped to sync with IPv6 version

=back

=head1 METHODS

=cut

package Net::CIDR::Lookup::Tie;

use strict;
use warnings;
use Carp;
use Net::CIDR::Lookup;

$Net::CIDR::Lookup::Tie::VERSION = sprintf "%d.%d", q$Revision: 0.41 $ =~ m/ (\d+) \. (\d+) /xg;

sub TIEHASH {   ## no critic (Subroutines::RequireArgUnpacking)
    my $class = shift;
    bless { tree => Net::CIDR::Lookup->new(@_) }, $class;
}

=head2 STORE

Stores a value under a given key

=cut

sub STORE { ## no critic (Subroutines::RequireArgUnpacking)
    my $self = shift;
    undef $self->{keys};
    if($_[0] =~ /-/) {
        $self->{tree}->add_range(@_);
    } else {
        $self->{tree}->add(@_);
    }
}

=head2 FETCH

Fetches the value stored under a given key

=cut

sub FETCH {
    my ($self, $key) = @_;
    $self->{tree}->lookup($key);
}
 
=head2 FIRSTKEY

Gets the first key in the hash. Used for iteration with each()

=cut

sub FIRSTKEY {
    my $self = shift;
    $self->_updkeys;
    each %{$self->{keys}};
}

=head2 NEXTKEY

Gets the next key from the hash. Used for iteration with each()

=cut

sub NEXTKEY {
    each %{shift->{keys}};
}

=head2 EXISTS

Tests if a key is in the hash. Also returns true for blocks or addresses
contained within a block that was actually stored.

=cut

sub EXISTS {
    my ($self, $key) = @_;
    $self->_updkeys;
    exists $self->{keys}{$key};
}

=head2 DELETE

Delete a key from the hash. Note that the same restrictions as for Net::CIDR::Lookup
regarding netblock splitting apply!

=cut

sub DELETE {
    carp('Deletions are not supported by tied ' . __PACKAGE__ . ' objects yet!');
}

=head2 CLEAR

Deletes all keys and their values.

=cut

sub CLEAR {
    my $self = shift;
    $self->{tree}->clear;
}

=head2 SCALAR

Returns the number of keys in the hash

=cut

sub SCALAR {
    my $self = shift;
    $self->_updkeys;
    scalar keys %{$self->{keys}};
}

=head2 _updkeys

Private method to update the internal key cache used for iteration

=cut

sub _updkeys {
    my $self = shift;

    if(defined $self->{keys}) {
        keys %{$self->{keys}};                  # Call in void context to reset
    } else {
        $self->{keys} = $self->{tree}->dump;    # Recreate hash
    }
}
1;


lib/Net/CIDR/Lookup.pm
=head1 NAME

Net::CIDR::Lookup

=head1 DESCRIPTION

This class implements a lookup table indexed by IPv4 networks or hosts.

=over 1

=item * Addresses are accepted in numeric form (integer with separate netbits argument),
as strings in CIDR notation or as IP address ranges

=item * Overlapping or adjacent networks are automatically coalesced if their
associated values are equal.

=item * The table is implemented as a binary tree so lookup and insertion take O(log n)
time.

=back

Methods generally return a true value on success and C<undef> on error. In the
latter case, an error message will be available in C<$Net::CIDR::Lookup::errstr>

=head1 SYNOPSIS

  use Net::CIDR::Lookup;

  $cidr = Net::CIDR::Lookup->new;
  $cidr->add("192.168.42.0/24",1);    # Add first network, value 1
  $cidr->add_num(167772448,27,2);     # 10.0.1.32/27 => 2
  $cidr->add("192.168.43.0/24",1);    # Automatic coalescing to a /23
  $cidr->add("192.168.41.0/24",2);    # Stays separate due to different value
  $cidr->add("192.168.42.128/25",2);  # Error: overlaps with different value

  $h = $cidr->dump;                   # Convert tree to a hash

  print "$k => $v\n" while(($k,$v) = each %$h);

  # Output (order may vary):
  # 192.168.42.0/23 => 1
  # 10.0.1.32/27 => 2
  # 192.168.41.0/24 => 2

  $cidr->walk(sub {
      my ($addr, $bits, $val) = @_;
      print join('.', unpack 'C*', pack 'N', $addr), "/$bits => $val\n"
    }
  );

  # Output (fixed order):
  # 10.0.1.32/27 => 2
  # 192.168.41.0/24 => 2
  # 192.168.42.0/23 => 1

  $cidr->clear;                                 # Remove all entries
  $cidr->add_range('1.2.3.11 - 1.2.4.234', 42); # Add a range of addresses, automatically split into CIDR blocks
  $h = $cidr->dump;
  print "$k => $v\n" while(($k,$v) = each %$h);

  # Output (order may vary):
  # 1.2.4.128/26 => 42
  # 1.2.3.32/27 => 42
  # 1.2.3.64/26 => 42
  # 1.2.4.234/32 => 42
  # 1.2.4.0/25 => 42
  # 1.2.3.12/30 => 42
  # 1.2.3.128/25 => 42
  # 1.2.3.16/28 => 42
  # 1.2.4.224/29 => 42
  # 1.2.4.232/31 => 42
  # 1.2.3.11/32 => 42
  # 1.2.4.192/27 => 42

=head1 HISTORY

=over 1

=item v0.3 First CPAN release

=item v0.3.1 

=over 1

=item * Replaced the simplistic list-based CIDR block splitting function with bit-fiddling for about a threefold speedup of C<add_num_range> and slightly less in C<add_range>.

=item * Recursive merging-up up of blocks during C<add_*> works now. If e.g.
you had a /24 and an adjacent /25 net with the same value before, adding a new
/25 would have merged the new block with the existing /25, resulting in two
adjacent /24s with the same value because only single-level merging was
possible. Now the two will be merged to a single /23.

=item * Removed some redundant tests and added new ones.

=item * Removed some leftover debug code.

=item * Some small fixes/improvements like stricter range checking in C<add_range>

=back

=item v0.4 Version bump for inclusion of the IPv6 version

=item v0.41 Version bumped to sync with IPv6 version

=back

=head1 METHODS

=cut

package Net::CIDR::Lookup;
use strict;
use warnings;
use Carp;

our $VERSION = sprintf "%d.%d", q$Revision: 0.41$ =~ m/ (\d+) \. (\d+) /xg;
our $errstr  = undef;

=head2 new

Arguments: none

Return Value: new object

=cut

sub new { bless [], shift }

=head2 add

Arguments: C<$cidr>, C<$value>

Return Value: true for successful completion, C<undef> otherwise

Adds VALUE to the tree under the key CIDR. CIDR must be a string containing an
IPv4 address followed by a slash and a number of network bits. Bits to the
right of this mask will be ignored.

=cut

sub add {
	my ($self, $cidr, $val) = @_;

    unless(defined $val) {
        $__PACKAGE__::errstr = "can't store an undef";
        return;
    }
	my ($net, $bits) = $cidr =~ m{ ^ ([.[:digit:]]+) / (\d+) $ }ox;
    unless(defined $net and defined $bits) {
        $__PACKAGE__::errstr = 'full dotted-quad/netbits notation required';
        return;
    }
    my $intnet = _dq2int($net) or return;
	$self->_add($intnet,$bits,$val);
}

=head2 add_range

Arguments: C<$range>, C<$value>

Return Value: true for successful completion, C<undef> otherwise

Adds VALUE to the tree for each address included in RANGE which must be a
hyphenated range of IP addresses in dotted-quad format (e.g.
"192.168.0.150-192.168.10.1") and with the first address being numerically
smaller the second. This range will be split up into as many CIDR blocks as
necessary (algorithm adapted from a script by Dr. Liviu Daia).

=cut

sub add_range {
    my ($self, $range, $val) = @_;

    unless(defined $val) {
        $__PACKAGE__::errstr = "can't store an undef";
        return;
    }

    my ($ip_start, $ip_end, $crud) = split /\s*-\s*/, $range;
    if(defined $crud or not defined $ip_end) {
        $__PACKAGE__::errstr = 'must have exactly one hyphen in range';
        return;
    }

    $ip_start = _dq2int($ip_start) or return;
    $ip_end   = _dq2int($ip_end)   or return;
    # This check will be repeated in add_num_range but we get more readable
    # errors here
    if($ip_start > $ip_end) {
        $__PACKAGE__::errstr = "start > end in range `$range'";
        return;
    }

    $self->add_num_range($ip_start, $ip_end, $val);
}

=head2 add_num

Arguments: C<$address>, C<$bits>, C<$value>

Return Value: true for successful completion, C<undef> otherwise

Like C<add()> but accepts address and bits as separate integer arguments
instead of a string.

=cut

sub add_num { ## no critic (Subroutines::RequireArgUnpacking)
    # my ($self,$ip,$bits,$val) = @_;
	# Just call the recursive adder for now but allow for changes in object
    # representation ($self != $n)
    unless(defined $_[3]) {
        $__PACKAGE__::errstr = "can't store an undef";
        return;
    }
	_add(@_);
}

=head2 add_num_range

Arguments: C<$start>, C<$end>, C<$value>

Return Value: true for successful completion, C<undef> otherwise

Like C<add_range()> but accepts addresses as separate integer arguments instead
of a range string.

=cut

sub add_num_range {
    my ($self, $start, $end, $val) = @_;

    if($start > $end) {
        $__PACKAGE__::errstr = "start > end in range $start--$end";
        return;
    }

    my @chunks;
    _do_chunk(\@chunks, $start, $end, 31, 0);
    foreach(@chunks) {
        $self->add_num(@$_, $val) or return;   # Immediately fail on first problem
    }
    1;
}

=head2 lookup

Arguments: C<$address>

Return Value: value assoiated with this address or C<undef>

Looks up an address and returns the value associated with the network
containing it. So far there is no way to tell which network that is though.

=cut

sub lookup {
	my ($self, $addr) = @_;

    # Make sure there is no network spec tacked onto $addr
    $addr =~ s!/.*!!;
	my $ip = _dq2int($addr) or return;
	$self->_lookup($ip, 32);
}


=head2 lookup_num

Arguments: C<$address>

Return Value: value assoiated with this address or C<undef>

Like C<lookup()> but accepts the address in integer form.

=cut

sub lookup_num { _lookup(@_, 32) } ## no critic (Subroutines::RequireArgUnpacking)

=head2 dump

Arguments: none

Return Value: C<$hashref>

Returns a hash representation of the tree with keys being CIDR-style network
addresses.

=cut

sub dump {  ## no critic (Subroutines::ProhibitBuiltinHomonyms)
	my ($self) = @_;
	my %result;
	$self->_walk(0, 0, sub {
            my $net = _int2dq($_[0]) . '/' . $_[1];
            if(defined $result{$net}) {
                confess("internal error: network $net mapped to $result{$net} already!\n");
            } else {
                $result{$net} = $_[2];
            }
        }
    );
	\%result;
}

=head2 walk

Arguments: C<$coderef> to call for each tree entry. Callback arguments are:

=over 1

=item C<$address>

The network address in integer form

=item C<$bits>

The current CIDR block's number of network bits

=item C<$value>

The value associated with this block

=back

Return Value: nothing useful

=cut

sub walk { $_[0]->_walk(0, 0, $_[1]) } ## no critic (Subroutines::RequireArgUnpacking)


=head2 clear

Arguments: none

Return Value: nothing useful

Remove all entries from the tree.

=cut

sub clear {
    my $self = shift;
    undef @$self;
}

=head1 BUGS

=over 1

=item *

I didn't need deletions yet and deleting parts of a CIDR block is a bit more
complicated than anything this class does so far, so it's not implemented.

=item *

Storing an C<undef> value does not work and yields an error. This would be
relatively easy to fix at the cost of some memory so that's more a design
decision.

=item *

Using a package-global for error reporting was an incredibly stupid idea
initially. This will change in the next version.

=back

=head1 AUTHORS, COPYRIGHTS & LICENSE

Matthias Bethke <matthias@towiski.de>
while working for 1&1 Internet AG

Licensed unter the Artistic License 2.0

=head1 SEE ALSO

This module's methods are based loosely on those of C<Net::CIDR::Lite>

=cut

# Walk through a subtree and insert a network
sub _add {
	my ($node, $addr, $nbits, $val) = @_;
    my ($bit, $checksub);
    my @node_stack;

    DESCEND:
    while(1) {
	    $bit = ($addr & 0x80000000) >> 31;
        $addr <<= 1;

        if(__PACKAGE__ ne ref $node) {
            return 1 if($val eq $node); # Compatible entry (tried to add a subnet of one already in the tree)
            $__PACKAGE__::errstr = "incompatible entry, found `$node' trying to add `$val'";
            return;
        }
        last DESCEND unless --$nbits;
        if(defined $node->[$bit]) {
            $checksub = 1;
        } else {
            $node->[$bit] ||= bless([], __PACKAGE__);
            $checksub = 0;
        }
        push @node_stack, \$node->[$bit];
        $node = $node->[$bit];
    }
    
    if($checksub and __PACKAGE__ eq ref $node->[$bit]) {
        _add_check_subtree($node->[$bit], $val) or return;
    }

    $node->[$bit] = $val;

    # Take care of potential mergers into the previous node (if $node[0] == $node[1])
    if(not @node_stack and defined $node->[$bit ^ 1] and $node->[$bit ^ 1] eq $val) {
        $__PACKAGE__::errstr = 'merging two /1 blocks is not supported yet';
        return;
    }
    while(1) {
        $node = pop @node_stack // last;
        last unless(defined $$node->[0] and defined $$node->[1] and $$node->[0] eq $$node->[1]);
        $$node = $val;
    }
    1;
}

# Check an existing subtree for incompatible values. Returns false and sets the
# package-global error string if there was a problem.
sub _add_check_subtree {
    my ($root, $val) = @_;

    eval {
        $root->_walk(0, 0, sub {
                my $oldval = $_[2];
                $val == $oldval or die $oldval; ## no critic (ErrorHandling::RequireCarping)
            }
        );
        1;
    } or do {
        if($@) {
            $__PACKAGE__::errstr = "incompatible entry, found `$@' trying to add `$val'";
            return;
        }
    };
    return 1;
}

sub _lookup {
	my ($node, $addr) = @_;
    my $bit;

    while(1) {
        $bit = ($addr & 0x80000000) >> 31;
        defined $node->[$bit] or return;
        __PACKAGE__ ne ref $node->[$bit] and return $node->[$bit];
        $node = $node->[$bit];
        $addr <<= 1;
    }
}

# Dotted-quad to integer
sub _dq2int { ## no critic (Subroutines::RequireArgUnpacking)
	my @oct = split /\./, $_[0];
	unless(4 == @oct) {
        $__PACKAGE__::errstr = "address must be in dotted-quad form, is `$_[0]'";
        return;
    }
	my $ip = 0;
    foreach(@oct) {
        if($_ > 255 or $_ < 0) {
            $__PACKAGE__::errstr = "invalid component `$_' in address `$_[0]'";
            return;
        }
        $ip = $ip<<8 | $_;
    }
	$ip;
}

# Convert an IP address in integer format to dotted-quad
sub _int2dq { join '.', unpack 'C*', pack 'N', shift }

# Convert a CIDR block ($addr, $bits) into a range of addresses ($lo, $hi)
# sub _cidr2rng { ( $_[0], $_[0] | ((1 << $_[1]) - 1) ) }

# Walk the tree in depth-first LTR order
sub _walk {
	my ($node, $addr, $bits, $cb) = @_;
	my ($l, $r);
    my @node_stack = ($bits, $node);
    #print "================== WALK ==================: ", join(':',caller),"\n"; 
    while(defined($node = pop @node_stack)) {
        $bits    = pop @node_stack;
        #print "LOOP: stack size ".@node_stack."\n";
        if(__PACKAGE__ eq ref $node) {
            ($l, $r) = @$node;
            #printf "Popped l=%s r=%s, bits=%d\n", ($l//'<undef>'), ($r//'<undef>'), $bits;
            ++$bits;

            # Check left side
            if(__PACKAGE__ eq ref $l) {
                #print "L: pushing node=$l, bits=$bits\n";
                defined $r and push @node_stack, ($bits, $r);
                push @node_stack, ($bits, $l);
                #printf "L: addr=%032b (%s)\n", $addr, _int2dq($addr);
                next; # Short-circuit back to loop w/o checking $r!
            } else {
                #defined $l and printf "L: CALLBACK (%s/%d) => %s\n", _int2dq($addr), $bits, $l;
                defined $l and $cb->($addr, $bits, $l);
            }
        } else {
            # There was a right-side leaf node on the stack that will end up in
            # the "else" branch below
            #print "Found leftover right leaf $node\n";
            $r = $node;
        }

        # Check right side
        if(__PACKAGE__ eq ref $r) {
            #print "R: pushing node=$r, bits=$bits\n";
            push @node_stack, ($bits, $r);
            $addr |= 1 << 32-$bits;
            #printf "R: addr=%032b (%s)\n", $addr, _int2dq($addr);
        } else {
            #defined $r and printf "R: CALLBACK (%s/%d) => %s\n", _int2dq($addr | 1 << 32-$bits), $bits, $r;
            defined $r and $cb->($addr | 1 << 32-$bits, $bits, $r);
        }
    }
}

# Split a chunk into a minimal number of CIDR blocks.
sub _do_chunk {
    my ($chunks, $start, $end, $ix1, $ix2) = @_;
    my ($prefix, $xor);

    # Find common prefix.  After that, the bit indicated by $ix1 is 0 for $start
    # and 1 for $end. A split a this point guarantees the longest suffix.
    $xor = $start ^ $end;
    --$ix1 until($xor & 1 << $ix1 or -1 == $ix1);
    $prefix = $start & ~((1 << ($ix1+1)) - 1);

    $ix2++ while($ix2 <= $ix1
            and not ($start & 1 << $ix2)
            and     ($end   & 1 << $ix2));

    # Split if $fbits and $lbits disagree on the length of the chunk.
    if ($ix2 <= $ix1) {
        _do_chunk($chunks, $start,              $prefix | ((1<<$ix1) - 1), $ix1, $ix2);
        _do_chunk($chunks, $prefix | (1<<$ix1), $end,                      $ix1, $ix2);
    } else {
        push @$chunks, [ $prefix, 31-$ix1 ];
    }
}

1;
